server/src/weaverbird/pipeline/formula_ast/types.py

```diff
@@ -0,0 +1,72 @@
+from ast import literal_eval
+from enum import Enum
+from typing import Union
+
+from pydantic import BaseModel
+
+
+class Operator(str, Enum):
+    ADD = '+'
+    SUB = '-'
+    MUL = '*'
+    DIV = '/'
+    MOD = '%'
+
+
+Constant = Union[int, bool, float, str]
+
+
+class ColumnName(BaseModel):
+    name: str
+    alias: str
+
+
+Expression = Union['Operation', ColumnName, Constant]
+
+
+# Dataclasses do not supported recursive types for now
```

Comment:
Maybe it's a bit overkill to bring pydantic for this? (we don't need validation)

Reply:
I agree that Pydantic is overkill (especially since I had to use `smart_union`) but I haven't been able to make recursive types work (at least not without some hacks) with dataclasses :confused: 

OTOH, most of our formulas are very simple, it's not like we'll build ASTs with several hundred pydantic model instances, so I don't think the performance overhead here is important compared to execution time. And we'd need some profiling to make sure this is really slower than regexes :smile: .

Maybe it could be implemented with typed NamedTuples though, I'll have a look into that !